ipfssync/subpub/discovery.go

```diff
@@ -0,0 +1,185 @@
+			if err != nil {
+				log.Fatal(err)
+			}
+			if peerinfo == nil {
+				continue // nothing to do
+			}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				log.Debugf("trying %s", *peerinfo)
+				if err := ps.Host.Connect(ctx, *peerinfo); err != nil {
+					log.Debug(err)
+				} else {
+					log.Infof("connection established with bootstrap node: %s", peerinfo)
+				}
+			}()
+		}
+		wg.Wait()
+	}
+
+	// We use a rendezvous point "meet me here" to announce our location.
+	// This is like telling your friends to meet you at the Eiffel Tower.
+	log.Infof("advertising myself periodically in topic %s", ps.Topic)
+	ps.routing = discovery.NewRoutingDiscovery(ps.dht)
+	discovery.Advertise(ctx, ps.routing, ps.Topic)
+
+	// Discover new peers periodically
+	go func() {
+		for {
+			select {
+			case <-ps.close:
+				return
+			default:
+				pctx, cancel := context.WithTimeout(ctx, ps.DiscoveryPeriod)
+				ps.discover(pctx)
+				cancel()
+				time.Sleep(ps.DiscoveryPeriod)
+			}
+		}
+	}()
+}
+
+func (ps *SubPub) discover(ctx context.Context) {
+	// Now, look for others who have announced.
+	// This is like your friend telling you the location to meet you.
+	log.Debugf("looking for peers in topic %s", ps.Topic)
+	peerChan, err := ps.routing.FindPeers(ctx, ps.Topic,
+		corediscovery.Limit(4*ps.MaxDHTpeers))
+	if err != nil {
+		log.Fatal(err)
+	}
+	for peer := range peerChan {
+		select {
+		case <-ps.close:
+			return
+		default:
+			// continues below
+		}
+		if peer.ID == ps.Host.ID() {
+			continue // this is us; skip
+		}
+
+		if ps.connectedPeer(peer.ID) {
+			continue
+		}
+		// new peer; let's connect to it
+		if err := peer.ID.Validate(); err == nil {
+			stream, err := ps.Host.NewStream(context.Background(), peer.ID, protocol.ID(ps.Topic))
+			if err != nil {
+				// Since this error is pretty common in p2p networks.
+				continue
+			}
+			log.Debugf("found peer %s: %v", peer.ID.Pretty(), peer.Addrs)
+			ps.Host.ConnManager().Protect(peer.ID, "discoveredPeer")
+			ps.handleStream(stream)
+		}
+	}
+}
+
+// connectedPeer returns true if the peer has some stream
+func (ps *SubPub) connectedPeer(pid libpeer.ID) bool {
+	for _, conn := range ps.Host.Network().ConnsToPeer(pid) {
+		if len(conn.GetStreams()) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// AddPeer creates a new libp2p peer connection (transport layer)
+func (ps *SubPub) AddPeer(peer string) error {
+	m, err := multiaddr.NewMultiaddr(peer)
+	if err != nil {
+		return err
+	}
+	ai, err := libpeer.AddrInfoFromP2pAddr(m)
+	if err != nil {
+		return err
+	}
+	ps.Host.ConnManager().Protect(ai.ID, "customPeer")
+	return ps.Host.Connect(context.Background(), *ai)
+}
+
+func (ps *SubPub) SetBootnodes(bootnodes []string) {
+	ps.BootNodes = bootnodes
+}
+
+func parseMultiaddress(maddress []string) (ma []multiaddr.Multiaddr) {
+	for _, m := range maddress {
+		mad, err := multiaddr.NewMultiaddr(m)
+		if err != nil {
+			log.Warn(err)
```

Comment:
sounds like this should be an error return - otherwise we'll happily ignore an invalid bootnode.